`control_statement` triggers on trailing closers requiring braces.

### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

SwiftLint rule `control_statement` triggers on statements that require braces due to usage of trailing closures.

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint test.swift
Linting Swift files at paths test.swift
Linting 'test.swift' (1/1)
/Users/hubstaff/Developer/hubstaff/ios/test.swift:4:5: warning: Control Statement Violation: `if`, `for`, `guard`, `switch`, `while`, and `catch` statements shouldn't unnecessarily wrap their conditionals or arguments in parentheses (control_statement)
Done linting! Found 1 violation, 0 serious in 1 file.
```

### Environment

* SwiftLint version: 0.52.4
* Installation method used: Homebrew
* Xcode Version 15.0 beta 4 (15A5195m)
* swift-driver version: 1.85 Apple Swift version 5.9 (swiftlang-5.9.0.120.7 clang-1500.0.34.3)
* Do you have a sample that shows the issue? Run `echo "[string here]" | swiftlint lint --no-cache --use-stdin --enable-all-rules`
  to quickly test if your example is really demonstrating the issue. If your example is more
  complex, you can use `swiftlint lint --path [file here] --no-cache --enable-all-rules`.

```swift
func foo() {
    var vm_info = task_vm_info()
    count = mach_msg_type_number_t(MemoryLayout<task_vm_info>.size) / 4
    if (withUnsafeMutablePointer(to: &vm_info) {
        $0.withMemoryRebound(to: task_info_t.self.Pointee, capacity: 1) {
            task_info(mach_task_self_, task_flavor_t(TASK_VM_INFO), $0, &count)
        }
    }) == KERN_SUCCESS {
        sample.memory_physical_used = vm_info.phys_footprint
    }
}
```

```bash
$ swiftlint lint --no-cache --use-stdin --enable-all-rules < test.swift
warning: The `inert_defer` rule is now deprecated and will be completely removed in a future release.
warning: The `unused_capture_list` rule is now deprecated and will be completely removed in a future release.
warning: The `anyobject_protocol` rule is now deprecated and will be completely removed in a future release.
<nopath>:5:9: warning: Anonymous Argument in Multiline Closure Violation: Use named arguments in multiline closures (anonymous_argument_in_multiline_closure)
<nopath>:6:69: warning: Anonymous Argument in Multiline Closure Violation: Use named arguments in multiline closures (anonymous_argument_in_multiline_closure)
<nopath>:4:5: warning: Control Statement Violation: `if`, `for`, `guard`, `switch`, `while`, and `catch` statements shouldn't unnecessarily wrap their conditionals or arguments in parentheses (control_statement)
<nopath>:1:1: warning: Explicit ACL Violation: All declarations should specify Access Control Level keywords explicitly (explicit_acl)
<nopath>:1:1: warning: Explicit Top Level ACL Violation: Top-level declarations should specify Access Control Level keywords explicitly (explicit_top_level_acl)
<nopath>:2:9: warning: Explicit Type Interface Violation: Properties should have a type interface (explicit_type_interface)
<nopath>:2:5: error: Identifier Name Violation: Variable name 'vm_info' should only contain alphanumeric and other allowed characters (identifier_name)
<nopath>:3:71: warning: No Magic Numbers Violation: Magic numbers should be replaced by named constants (no_magic_numbers)
Done linting! Found 8 violations, 1 serious in 1 file.
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -16,6 +16,9 @@
   [chrisngabp](https://github.com/chrisngabp)
   [5103](https://github.com/realm/SwiftLint/issues/5103)
 
+* Rewrite `control_statement` rule using SwiftSyntax.  
+  [SimplyDanny](https://github.com/SimplyDanny)
+
 #### Bug Fixes
 
 * Fix bug in `prefer_self_in_static_references` rule that triggered on

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -21,6 +21,11 @@
 
 #### Bug Fixes
 
+* Fix false positive in `control_statement` rule that triggered on conditions
+  with trailing closures where parentheses are recommended by the compiler.  
+  [SimplyDanny](https://github.com/SimplyDanny)
+  [#5135](https://github.com/realm/SwiftLint/issues/5135)
+
 * Fix bug in `prefer_self_in_static_references` rule that triggered on
   initializers of computed properties in classes when the property had an
   accessor block.  

--- a/Source/SwiftLintBuiltInRules/Rules/Lint/MissingDocsRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Lint/MissingDocsRule.swift
@@ -28,7 +28,7 @@ private extension SwiftLintFile {
         }
 
         guard let kind = dictionary.declarationKind,
-            (!SwiftDeclarationKind.extensionKinds.contains(kind) || !excludesExtensions),
+            !SwiftDeclarationKind.extensionKinds.contains(kind) || !excludesExtensions,
             case let isDeinit = kind == .functionMethodInstance && dictionary.name == "deinit",
             !isDeinit,
             let offset = dictionary.offset,

--- a/Source/SwiftLintBuiltInRules/Rules/Style/ControlStatementRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Style/ControlStatementRule.swift
@@ -187,7 +187,7 @@ private final class Rewriter: SyntaxRewriter, ViolationsSyntaxRewriter {
     }
 }
 
-private class ClosureExprFinder: SyntaxTransformVisitor {
+private class TrailingClosureFinder: SyntaxTransformVisitor {
     func visitAny(_ node: Syntax) -> Bool {
         false
     }
@@ -204,7 +204,7 @@ private class ClosureExprFinder: SyntaxTransformVisitor {
 private extension ExprSyntax {
     var unwrapped: ExprSyntax? {
         if let expr = self.as(TupleExprSyntax.self)?.elementList.onlyElement?.expression {
-            return ClosureExprFinder().visit(expr) ? nil : expr
+            return TrailingClosureFinder().visit(expr) ? nil : expr
         }
         return nil
     }

--- a/Source/SwiftLintBuiltInRules/Rules/Style/ControlStatementRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Style/ControlStatementRule.swift
@@ -1,7 +1,6 @@
-import Foundation
-import SourceKittenFramework
+import SwiftSyntax
 
-struct ControlStatementRule: ConfigurationProviderRule, SubstitutionCorrectableRule {
+struct ControlStatementRule: ConfigurationProviderRule, SwiftSyntaxCorrectableRule {
     var configuration = SeverityConfiguration<Self>(.warning)
 
     static let description = RuleDescription(
@@ -12,145 +11,230 @@ struct ControlStatementRule: ConfigurationProviderRule, SubstitutionCorrectableR
             "conditionals or arguments in parentheses",
         kind: .style,
         nonTriggeringExamples: [
-            Example("if condition {\n"),
-            Example("if (a, b) == (0, 1) {\n"),
-            Example("if (a || b) && (c || d) {\n"),
-            Example("if (min...max).contains(value) {\n"),
-            Example("if renderGif(data) {\n"),
-            Example("renderGif(data)\n"),
-            Example("for item in collection {\n"),
-            Example("for (key, value) in dictionary {\n"),
-            Example("for (index, value) in enumerate(array) {\n"),
-            Example("for var index = 0; index < 42; index++ {\n"),
-            Example("guard condition else {\n"),
-            Example("while condition {\n"),
-            Example("} while condition {\n"),
-            Example("do { ; } while condition {\n"),
-            Example("switch foo {\n"),
-            Example("do {\n} catch let error as NSError {\n}"),
+            Example("if condition {}"),
+            Example("if (a, b) == (0, 1) {}"),
+            Example("if (a || b) && (c || d) {}"),
+            Example("if (min...max).contains(value) {}"),
+            Example("if renderGif(data) {}"),
+            Example("renderGif(data)"),
+            Example("guard condition else {}"),
+            Example("while condition {}"),
+            Example("do {} while condition {}"),
+            Example("do { ; } while condition {}"),
+            Example("switch foo {}"),
+            Example("do {} catch let error as NSError {}"),
             Example("foo().catch(all: true) {}"),
-            Example("if max(a, b) < c {\n"),
-            Example("switch (lhs, rhs) {\n")
+            Example("if max(a, b) < c {}"),
+            Example("switch (lhs, rhs) {}")
         ],
         triggeringExamples: [
-            Example("↓if (condition) {\n"),
-            Example("↓if(condition) {\n"),
-            Example("↓if (condition == endIndex) {\n"),
-            Example("↓if ((a || b) && (c || d)) {\n"),
-            Example("↓if ((min...max).contains(value)) {\n"),
-            Example("↓for (item in collection) {\n"),
-            Example("↓for (var index = 0; index < 42; index++) {\n"),
-            Example("↓for(item in collection) {\n"),
-            Example("↓for(var index = 0; index < 42; index++) {\n"),
-            Example("↓guard (condition) else {\n"),
-            Example("↓while (condition) {\n"),
-            Example("↓while(condition) {\n"),
-            Example("} ↓while (condition) {\n"),
-            Example("} ↓while(condition) {\n"),
-            Example("do { ; } ↓while(condition) {\n"),
-            Example("do { ; } ↓while (condition) {\n"),
-            Example("↓switch (foo) {\n"),
-            Example("do {\n} ↓catch(let error as NSError) {\n}"),
-            Example("↓if (max(a, b) < c) {\n")
+            Example("↓if (condition) {}"),
+            Example("↓if(condition) {}"),
+            Example("↓if (condition == endIndex) {}"),
+            Example("↓if ((a || b) && (c || d)) {}"),
+            Example("↓if ((min...max).contains(value)) {}"),
+            Example("↓guard (condition) else {}"),
+            Example("↓while (condition) {}"),
+            Example("↓while(condition) {}"),
+            Example("do { ; } ↓while(condition) {}"),
+            Example("do { ; } ↓while (condition) {}"),
+            Example("↓switch (foo) {}"),
+            Example("do {} ↓catch(let error as NSError) {}"),
+            Example("↓if (max(a, b) < c) {}")
         ],
         corrections: [
-            Example("↓if (condition) {\n"): Example("if condition {\n"),
-            Example("↓if(condition) {\n"): Example("if condition {\n"),
-            Example("↓if (condition == endIndex) {\n"): Example("if condition == endIndex {\n"),
-            Example("↓if ((a || b) && (c || d)) {\n"): Example("if (a || b) && (c || d) {\n"),
-            Example("↓if ((min...max).contains(value)) {\n"): Example("if (min...max).contains(value) {\n"),
-            Example("↓for (item in collection) {\n"): Example("for item in collection {\n"),
-            Example("↓for (var index = 0; index < 42; index++) {\n"):
-                Example("for var index = 0; index < 42; index++ {\n"),
-            Example("↓for(item in collection) {\n"): Example("for item in collection {\n"),
-            Example("↓for(var index = 0; index < 42; index++) {\n"):
-                Example("for var index = 0; index < 42; index++ {\n"),
-            Example("↓guard (condition) else {\n"): Example("guard condition else {\n"),
-            Example("↓while (condition) {\n"): Example("while condition {\n"),
-            Example("↓while(condition) {\n"): Example("while condition {\n"),
-            Example("} ↓while (condition) {\n"): Example("} while condition {\n"),
-            Example("} ↓while(condition) {\n"): Example("} while condition {\n"),
-            Example("do { ; } ↓while(condition) {\n"): Example("do { ; } while condition {\n"),
-            Example("do { ; } ↓while (condition) {\n"): Example("do { ; } while condition {\n"),
-            Example("↓switch (foo) {\n"): Example("switch foo {\n"),
-            Example("do {\n} ↓catch(let error as NSError) {\n}"): Example("do {\n} catch let error as NSError {\n}"),
-            Example("↓if (max(a, b) < c) {\n"): Example("if max(a, b) < c {\n")
+            Example("↓if (condition) {}"): Example("if condition {}"),
+            Example("↓if(condition) {}"): Example("if condition {}"),
+            Example("↓if (condition == endIndex) {}"): Example("if condition == endIndex {}"),
+            Example("↓if ((a || b) && (c || d)) {}"): Example("if (a || b) && (c || d) {}"),
+            Example("↓if ((min...max).contains(value)) {}"): Example("if (min...max).contains(value) {}"),
+            Example("↓guard (condition) else {}"): Example("guard condition else {}"),
+            Example("↓while (condition) {}"): Example("while condition {}"),
+            Example("↓while(condition) {}"): Example("while condition {}"),
+            Example("do {} ↓while (condition) {}"): Example("do {} while condition {}"),
+            Example("do {} ↓while(condition) {}"): Example("do {} while condition {}"),
+            Example("do { ; } ↓while(condition) {}"): Example("do { ; } while condition {}"),
+            Example("do { ; } ↓while (condition) {}"): Example("do { ; } while condition {}"),
+            Example("↓switch (foo) {}"): Example("switch foo {}"),
+            Example("do {} ↓catch(let error as NSError) {}"): Example("do {} catch let error as NSError {}"),
+            Example("↓if (max(a, b) < c) {}"): Example("if max(a, b) < c {}"),
+            Example("""
+            if (a),
+               ( b == 1 ) {}
+            """): Example("""
+                if a,
+                   b == 1 {}
+                """)
         ]
     )
 
-    func validate(file: SwiftLintFile) -> [StyleViolation] {
-        return violationRanges(in: file).map { match -> StyleViolation in
-            return StyleViolation(ruleDescription: Self.description,
-                                  severity: configuration.severity,
-                                  location: Location(file: file, characterOffset: match.location))
+    func makeVisitor(file: SwiftLintCore.SwiftLintFile) -> ViolationsSyntaxVisitor {
+        Visitor(viewMode: .sourceAccurate)
+    }
+
+    func makeRewriter(file: SwiftLintCore.SwiftLintFile) -> ViolationsSyntaxRewriter? {
+        Rewriter(
+            locationConverter: file.locationConverter,
+            disabledRegions: disabledRegions(file: file)
+        )
+    }
+}
+
+private final class Visitor: ViolationsSyntaxVisitor {
+    override var skippableDeclarations: [DeclSyntaxProtocol.Type] { [ProtocolDeclSyntax.self] }
+
+    override func visitPost(_ node: CatchClauseSyntax) {
+        if node.catchItems?.haveParens == true {
+            violations.append(node.positionAfterSkippingLeadingTrivia)
         }
     }
 
-    func violationRanges(in file: SwiftLintFile) -> [NSRange] {
-        let statements = ["if", "for", "guard", "switch", "while", "catch"]
-        let statementPatterns: [String] = statements.map { statement -> String in
-            let isGuard = statement == "guard"
-            let isSwitch = statement == "switch"
-            let elsePattern = isGuard ? "else\\s*" : ""
-            let clausePattern = isSwitch ? "[^,{]*" : "[^{]*"
-            return "\(statement)\\s*\\(\(clausePattern)\\)\\s*\(elsePattern)\\{"
-        }
-        return statementPatterns.flatMap { pattern -> [NSRange] in
-            return file.match(pattern: pattern)
-                .filter { match, syntaxKinds -> Bool in
-                    let matchString = file.contents.substring(from: match.location, length: match.length)
-                    return !isFalsePositive(matchString, syntaxKind: syntaxKinds.first)
-                }
-                .map { $0.0 }
-                .filter { match -> Bool in
-                    let contents = file.stringView
-                    guard let byteOffset = contents.NSRangeToByteRange(start: match.location, length: 1)?.location,
-                        let outerKind = file.structureDictionary.structures(forByteOffset: byteOffset).last else {
-                            return true
-                    }
-
-                    return outerKind.expressionKind != .call
-                }
+    override func visitPost(_ node: GuardStmtSyntax) {
+        if node.conditions.haveParens {
+            violations.append(node.positionAfterSkippingLeadingTrivia)
         }
     }
 
-    func substitution(for violationRange: NSRange, in file: SwiftLintFile) -> (NSRange, String)? {
-        var violationString = file.stringView.substring(with: violationRange)
-        if violationString.contains("(") && violationString.contains(")") {
-            if let openingIndex = violationString.firstIndex(of: "(") {
-                let replacement = violationString[violationString.index(before: openingIndex)] == " " ? "" : " "
-                violationString.replaceSubrange(openingIndex...openingIndex, with: replacement)
-            }
-            if let closingIndex = violationString.lastIndex(of: ")" as Character) {
-                let replacement = violationString[violationString.index(after: closingIndex)] == " " ? "" : " "
-                violationString.replaceSubrange(closingIndex...closingIndex, with: replacement)
-            }
+    override func visitPost(_ node: IfExprSyntax) {
+        if node.conditions.haveParens {
+            violations.append(node.positionAfterSkippingLeadingTrivia)
         }
-        return (violationRange, violationString)
     }
 
-    private func isFalsePositive(_ content: String, syntaxKind: SyntaxKind?) -> Bool {
-        if syntaxKind != .keyword {
-            return true
+    override func visitPost(_ node: SwitchExprSyntax) {
+        if node.expression.tupleElement != nil {
+            violations.append(node.positionAfterSkippingLeadingTrivia)
         }
+    }
 
-        guard let lastClosingParenthesePosition = content.lastIndex(of: ")") else {
-            return false
+    override func visitPost(_ node: WhileStmtSyntax) {
+        if node.conditions.haveParens {
+            violations.append(node.positionAfterSkippingLeadingTrivia)
         }
+    }
+}
 
-        var depth = 0
-        var index = 0
-        for char in content {
-            if char == ")" {
-                if index != lastClosingParenthesePosition && depth == 1 {
-                    return true
-                }
-                depth -= 1
-            } else if char == "(" {
-                depth += 1
+private final class Rewriter: SyntaxRewriter, ViolationsSyntaxRewriter {
+    private(set) var correctionPositions: [AbsolutePosition] = []
+    let locationConverter: SourceLocationConverter
+    let disabledRegions: [SourceRange]
+
+    init(locationConverter: SourceLocationConverter, disabledRegions: [SourceRange]) {
+        self.locationConverter = locationConverter
+        self.disabledRegions = disabledRegions
+    }
+
+    override func visit(_ node: CatchClauseSyntax) -> CatchClauseSyntax {
+        guard !node.isContainedIn(regions: disabledRegions, locationConverter: locationConverter),
+              let items = node.catchItems, items.haveParens == true else {
+            return super.visit(node)
+        }
+        correctionPositions.append(node.positionAfterSkippingLeadingTrivia)
+        let node = node
+            .with(\.catchKeyword, node.catchKeyword.with(\.trailingTrivia, .space))
+            .with(\.catchItems, items.withoutParens)
+        return super.visit(node)
+    }
+
+    override func visit(_ node: GuardStmtSyntax) -> StmtSyntax {
+        guard !node.isContainedIn(regions: disabledRegions, locationConverter: locationConverter),
+              node.conditions.haveParens else {
+            return super.visit(node)
+        }
+        correctionPositions.append(node.positionAfterSkippingLeadingTrivia)
+        let node = node
+            .with(\.guardKeyword, node.guardKeyword.with(\.trailingTrivia, .space))
+            .with(\.conditions, node.conditions.withoutParens)
+        return super.visit(node)
+    }
+
+    override func visit(_ node: IfExprSyntax) -> ExprSyntax {
+        guard !node.isContainedIn(regions: disabledRegions, locationConverter: locationConverter),
+              node.conditions.haveParens else {
+            return super.visit(node)
+        }
+        correctionPositions.append(node.positionAfterSkippingLeadingTrivia)
+        let node = node
+            .with(\.ifKeyword, node.ifKeyword.with(\.trailingTrivia, .space))
+            .with(\.conditions, node.conditions.withoutParens)
+        return super.visit(node)
+    }
+
+    override func visit(_ node: SwitchExprSyntax) -> ExprSyntax {
+        guard !node.isContainedIn(regions: disabledRegions, locationConverter: locationConverter),
+              let tupleElement = node.expression.tupleElement else {
+            return super.visit(node)
+        }
+        correctionPositions.append(node.positionAfterSkippingLeadingTrivia)
+        let node = node
+            .with(\.switchKeyword, node.switchKeyword.with(\.trailingTrivia, .space))
+            .with(\.expression, tupleElement.with(\.trailingTrivia, .space))
+        return super.visit(node)
+    }
+
+    override func visit(_ node: WhileStmtSyntax) -> StmtSyntax {
+        guard !node.isContainedIn(regions: disabledRegions, locationConverter: locationConverter),
+              node.conditions.haveParens else {
+            return super.visit(node)
+        }
+        correctionPositions.append(node.positionAfterSkippingLeadingTrivia)
+        let node = node
+            .with(\.whileKeyword, node.whileKeyword.with(\.trailingTrivia, .space))
+            .with(\.conditions, node.conditions.withoutParens)
+        return super.visit(node)
+    }
+}
+
+private extension ExprSyntax {
+    var tupleElement: ExprSyntax? {
+        self.as(TupleExprSyntax.self)?.elementList.onlyElement?.expression
+    }
+}
+
+private extension ConditionElementListSyntax {
+    var haveParens: Bool {
+        contains { $0.condition.is(TupleExprSyntax.self) }
+    }
+
+    var withoutParens: Self {
+        let conditions = map { (element: ConditionElementSyntax) -> ConditionElementSyntax in
+            if let expression = element.condition.as(ExprSyntax.self)?.tupleElement {
+                return element
+                    .with(\.condition, .expression(expression))
+                    .with(\.leadingTrivia, element.leadingTrivia)
+                    .with(\.trailingTrivia, element.trailingTrivia)
             }
-            index += 1
+            return element
         }
-        return false
+        return Self(conditions)
+            .with(\.leadingTrivia, leadingTrivia)
+            .with(\.trailingTrivia, trailingTrivia)
+    }
+}
+
+private extension CatchItemListSyntax {
+    var haveParens: Bool {
+        contains { $0.tupleElement != nil }
+    }
+
+    var withoutParens: Self {
+        let items = map { (item: CatchItemSyntax) -> CatchItemSyntax in
+            if let expression = item.tupleElement {
+                return item
+                    .with(\.pattern, PatternSyntax(ExpressionPatternSyntax(expression: expression)))
+                    .with(\.leadingTrivia, item.leadingTrivia)
+                    .with(\.trailingTrivia, item.trailingTrivia)
+            }
+            return item
+        }
+        return Self(items)
+            .with(\.leadingTrivia, leadingTrivia)
+            .with(\.trailingTrivia, trailingTrivia)
+    }
+}
+
+private extension CatchItemSyntax {
+    var tupleElement: ExprSyntax? {
+        pattern?.as(ExpressionPatternSyntax.self)?.expression.tupleElement
     }
 }

--- a/Source/SwiftLintBuiltInRules/Rules/Style/ControlStatementRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Style/ControlStatementRule.swift
@@ -25,7 +25,9 @@ struct ControlStatementRule: ConfigurationProviderRule, SwiftSyntaxCorrectableRu
             Example("do {} catch let error as NSError {}"),
             Example("foo().catch(all: true) {}"),
             Example("if max(a, b) < c {}"),
-            Example("switch (lhs, rhs) {}")
+            Example("switch (lhs, rhs) {}"),
+            Example("if (f() { g() {} }) {}"),
+            Example("if (a + f() {} == 1) {}")
         ],
         triggeringExamples: [
             Example("↓if (condition) {}"),
@@ -84,31 +86,31 @@ private final class Visitor: ViolationsSyntaxVisitor {
     override var skippableDeclarations: [DeclSyntaxProtocol.Type] { [ProtocolDeclSyntax.self] }
 
     override func visitPost(_ node: CatchClauseSyntax) {
-        if node.catchItems?.haveParens == true {
+        if node.catchItems?.containSuperfluousParens == true {
             violations.append(node.positionAfterSkippingLeadingTrivia)
         }
     }
 
     override func visitPost(_ node: GuardStmtSyntax) {
-        if node.conditions.haveParens {
+        if node.conditions.containSuperfluousParens {
             violations.append(node.positionAfterSkippingLeadingTrivia)
         }
     }
 
     override func visitPost(_ node: IfExprSyntax) {
-        if node.conditions.haveParens {
+        if node.conditions.containSuperfluousParens {
             violations.append(node.positionAfterSkippingLeadingTrivia)
         }
     }
 
     override func visitPost(_ node: SwitchExprSyntax) {
-        if node.expression.tupleElement != nil {
+        if node.expression.unwrapped != nil {
             violations.append(node.positionAfterSkippingLeadingTrivia)
         }
     }
 
     override func visitPost(_ node: WhileStmtSyntax) {
-        if node.conditions.haveParens {
+        if node.conditions.containSuperfluousParens {
             violations.append(node.positionAfterSkippingLeadingTrivia)
         }
     }
@@ -126,7 +128,7 @@ private final class Rewriter: SyntaxRewriter, ViolationsSyntaxRewriter {
 
     override func visit(_ node: CatchClauseSyntax) -> CatchClauseSyntax {
         guard !node.isContainedIn(regions: disabledRegions, locationConverter: locationConverter),
-              let items = node.catchItems, items.haveParens == true else {
+              let items = node.catchItems, items.containSuperfluousParens == true else {
             return super.visit(node)
         }
         correctionPositions.append(node.positionAfterSkippingLeadingTrivia)
@@ -138,7 +140,7 @@ private final class Rewriter: SyntaxRewriter, ViolationsSyntaxRewriter {
 
     override func visit(_ node: GuardStmtSyntax) -> StmtSyntax {
         guard !node.isContainedIn(regions: disabledRegions, locationConverter: locationConverter),
-              node.conditions.haveParens else {
+              node.conditions.containSuperfluousParens else {
             return super.visit(node)
         }
         correctionPositions.append(node.positionAfterSkippingLeadingTrivia)
@@ -150,7 +152,7 @@ private final class Rewriter: SyntaxRewriter, ViolationsSyntaxRewriter {
 
     override func visit(_ node: IfExprSyntax) -> ExprSyntax {
         guard !node.isContainedIn(regions: disabledRegions, locationConverter: locationConverter),
-              node.conditions.haveParens else {
+              node.conditions.containSuperfluousParens else {
             return super.visit(node)
         }
         correctionPositions.append(node.positionAfterSkippingLeadingTrivia)
@@ -162,7 +164,7 @@ private final class Rewriter: SyntaxRewriter, ViolationsSyntaxRewriter {
 
     override func visit(_ node: SwitchExprSyntax) -> ExprSyntax {
         guard !node.isContainedIn(regions: disabledRegions, locationConverter: locationConverter),
-              let tupleElement = node.expression.tupleElement else {
+              let tupleElement = node.expression.unwrapped else {
             return super.visit(node)
         }
         correctionPositions.append(node.positionAfterSkippingLeadingTrivia)
@@ -174,7 +176,7 @@ private final class Rewriter: SyntaxRewriter, ViolationsSyntaxRewriter {
 
     override func visit(_ node: WhileStmtSyntax) -> StmtSyntax {
         guard !node.isContainedIn(regions: disabledRegions, locationConverter: locationConverter),
-              node.conditions.haveParens else {
+              node.conditions.containSuperfluousParens else {
             return super.visit(node)
         }
         correctionPositions.append(node.positionAfterSkippingLeadingTrivia)
@@ -185,20 +187,42 @@ private final class Rewriter: SyntaxRewriter, ViolationsSyntaxRewriter {
     }
 }
 
+private class ClosureExprFinder: SyntaxTransformVisitor {
+    func visitAny(_ node: Syntax) -> Bool {
+        false
+    }
+
+    func visit(_ node: FunctionCallExprSyntax) -> Bool {
+        node.trailingClosure != nil
+    }
+
+    func visit(_ node: SequenceExprSyntax) -> Bool {
+        node.elements.contains(where: visit)
+    }
+}
+
 private extension ExprSyntax {
-    var tupleElement: ExprSyntax? {
-        self.as(TupleExprSyntax.self)?.elementList.onlyElement?.expression
+    var unwrapped: ExprSyntax? {
+        if let expr = self.as(TupleExprSyntax.self)?.elementList.onlyElement?.expression {
+            return ClosureExprFinder().visit(expr) ? nil : expr
+        }
+        return nil
     }
 }
 
 private extension ConditionElementListSyntax {
-    var haveParens: Bool {
-        contains { $0.condition.is(TupleExprSyntax.self) }
+    var containSuperfluousParens: Bool {
+        contains {
+            if case let .expression(wrapped) = $0.condition {
+                return wrapped.unwrapped != nil
+            }
+            return false
+        }
     }
 
     var withoutParens: Self {
         let conditions = map { (element: ConditionElementSyntax) -> ConditionElementSyntax in
-            if let expression = element.condition.as(ExprSyntax.self)?.tupleElement {
+            if let expression = element.condition.as(ExprSyntax.self)?.unwrapped {
                 return element
                     .with(\.condition, .expression(expression))
                     .with(\.leadingTrivia, element.leadingTrivia)
@@ -213,13 +237,13 @@ private extension ConditionElementListSyntax {
 }
 
 private extension CatchItemListSyntax {
-    var haveParens: Bool {
-        contains { $0.tupleElement != nil }
+    var containSuperfluousParens: Bool {
+        contains { $0.unwrapped != nil }
     }
 
     var withoutParens: Self {
         let items = map { (item: CatchItemSyntax) -> CatchItemSyntax in
-            if let expression = item.tupleElement {
+            if let expression = item.unwrapped {
                 return item
                     .with(\.pattern, PatternSyntax(ExpressionPatternSyntax(expression: expression)))
                     .with(\.leadingTrivia, item.leadingTrivia)
@@ -234,7 +258,7 @@ private extension CatchItemListSyntax {
 }
 
 private extension CatchItemSyntax {
-    var tupleElement: ExprSyntax? {
-        pattern?.as(ExpressionPatternSyntax.self)?.expression.tupleElement
+    var unwrapped: ExprSyntax? {
+        pattern?.as(ExpressionPatternSyntax.self)?.expression.unwrapped
     }
 }

--- a/Source/SwiftLintBuiltInRules/Rules/Style/ImplicitGetterRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Style/ImplicitGetterRule.swift
@@ -36,7 +36,7 @@ private final class ImplicitGetterRuleVisitor: ViolationsSyntaxVisitor {
               node.setAccessor == nil,
               getAccessor.effectSpecifiers == nil,
               getAccessor.modifier == nil,
-              (getAccessor.attributes == nil || getAccessor.attributes?.isEmpty == true),
+              getAccessor.attributes == nil || getAccessor.attributes?.isEmpty == true,
               getAccessor.body != nil else {
             return
         }

